File: TrainWeb.Domain/Domain/Booking.cs
using TrainWeb.Domain.Enum;

namespace TrainWeb.Domain.Domain
{
    public class Booking
    {
        public string? Id { get; }

        public string UserId { get; }
        public string TripId { get; }
        public string SeatId { get; }

        public string? TicketTypeId { get; } 

        public double Amount { get; set; }

        public BookingStatus Status { get; set; }
        public PaymentStatus PaymentStatus { get; set; }

        public string? PaymentId { get; set; }
        public string? TicketId { get; set; }
        public string? TicketStatus { get; set; }

        public DateTime CreatedAt { get; }
        public DateTime? ExpiresAt { get; }

        public Dictionary<string, object>? SeatSummary { get; set; }
        public Dictionary<string, object>? TripSummary { get; set; }

        public Booking(
            string? id,
            string userId,
            string tripId,
            string seatId,
            string? ticketTypeId,      
            double amount,
            BookingStatus status,
            PaymentStatus paymentStatus,
            string? paymentId,
            string? ticketId,
            string? ticketStatus,
            DateTime createdAt,
            DateTime? expiresAt
        )
        {
            Id = id;
            UserId = userId;
            TripId = tripId;
            SeatId = seatId;

            TicketTypeId = ticketTypeId; 

            Amount = amount;
            Status = status;
            PaymentStatus = paymentStatus;
            PaymentId = paymentId;
            TicketId = ticketId;
            TicketStatus = ticketStatus;
            CreatedAt = createdAt;
            ExpiresAt = expiresAt;
        }
    }
}

File: TrainWebUI/src/types/booking.ts
import { UserDto } from './user';
import { TripDto } from './trip';
import { SeatDto } from './seat';
import { TicketEntity } from './ticket';
import { PaymentStatus } from './payment';

export enum BookingStatus {
  Reserved = 0,
  Paid = 1,
  Cancelled = 2
}

export interface BookingEntity {
  id: string;
  userId: string;
  tripId: string;
  seatId: string;
  status: BookingStatus;
  createdAt: string;
}

export interface BookingDto {
  id?: string;
  userId?: string;
  tripId?: string;
  seatId?: string;
  ticketTypeId?: string;
  amount?: number;
  status?: BookingStatus | string;
  paymentStatus?: PaymentStatus | string;
  paymentId?: string | null;
  ticketId?: string | null;
  ticketStatus?: string | null;
  createdAt?: string;
  expiresAt?: string;
  seatSummary?: any;
  tripSummary?: any;
  user?: UserDto;
  trip?: TripDto;
  seat?: SeatDto;
  ticket?: TicketEntity; // Ticket reference for creating bookings
}

export const BOOKING_STATUS_LABELS: Record<BookingStatus, string> = {
  [BookingStatus.Reserved]: 'Đã đặt',
  [BookingStatus.Paid]: 'Đã thanh toán',
  [BookingStatus.Cancelled]: 'Đã hủy'
};

export function isBookingStatus(value: number): value is BookingStatus {
  return value >= 0 && value <= 2;
}

// Additional types for booking workflow
export interface Passenger {
  name: string;
  idNumber: string;
  phone: string;
}

export interface PassengerDto {
  name: string;
  idNumber: string;
  phone: string;
}

export interface SelectedSeat {
  seatId: string;
  seatNumber: string;
  type: string;
  price: number;
}
